### ISSUE TITLE
Investigate top-p performance

### ISSUE THREAD

**Issue by mattdangerw** [keras-team/keras-nlp#963]
Currently our top-p (nucleus) sampler has the worst performance of our entire sampling offering, likely because of the sort operation we need to do for our softmax probabilities. See [these rough numbers](https://github.com/keras-team/keras-nlp/commit/41af01200c5d9898f381477b6a67a87b6951f844#diff-576c27794bae284efc278336a16d0649aaf599973a64d4b65dc67bc99bb7f000) for example, where top-p is slower than beam, despite needing much less computation on the model itself.

We should see if there is anything we can do to speed up top-p sampling with XLA (e.g. a more XLA friendly sort op).

**Comment by mattdangerw** [comment]
One thing we could try here is adding a top-k heuristic cutoff, before applying top-p. This would mean you only need to sort a much smaller range of vocabulary terms to find the correct top-p cutoff point.

We could expose this as an option fairly easily. Less clear if we could expose some sort of defaults for this, but if we could it could give us a much more performant top-p out of the box.
